osu.Game/Rulesets/Scoring/HitEventExtensions.cs

```diff
@@ -65,6 +70,39 @@ public static class HitEventExtensions
             return timeOffsets.Average();
         }
 
-        public static bool AffectsUnstableRate(HitEvent e) => !(e.HitObject.HitWindows is HitWindows.EmptyHitWindows) && e.Result.IsHit();
+        public static bool AffectsUnstableRate(HitEvent e) => AffectsUnstableRate(e.HitObject, e.Result);
+        public static bool AffectsUnstableRate(HitObject hitObject, HitResult result) => hitObject.HitWindows != HitWindows.Empty && result.IsHit();
+
+        /// <summary>
+        /// Data type returned by <see cref="HitEventExtensions.CalculateUnstableRate"/> which allows efficient incremental processing.
+        /// </summary>
+        /// <remarks>
+        /// This should be passed back into future <see cref="HitEventExtensions.CalculateUnstableRate"/> calls as a parameter.
+        ///
+        /// The optimisations used here rely on hit events being a consecutive sequence from a single gameplay session.
+        /// When a new gameplay session is started, any existing results should be disposed.
+        /// </remarks>
+        public class UnstableRateCalculationResult
```

Comment:
My only query is why this isn't a `struct` with readonly fields.

Reply:
I will try a `struct`, but as a record the alloc overhead (cpu, not memory) outweighted the benefits, seemingly (see https://github.com/ppy/osu/pull/30868/commits/ea68d4b33abbd920e267001fb4785ae000031f54).

Reply:
If it has a significant overhead, I'd say leave it then.